### ISSUE TITLE
tf.grad() returns matrix of zeros

### ISSUE THREAD

**Issue by pxrrylson** [tensorflow/tfjs#4570]
I trained a CNN on an MNIST dataset, and planned to use it for inference with tf.js.

However, when checking the model's gradients with respect to one of the test images, I notice I keep getting a zero matrix. 

Does anybody know why this issue occur?
```
function getGradient(img, yTrue) {
    yTrue = tf.oneHot(tf.tensor1d([yTrue], 'int32'), 10);

    function f(x) {
        return tf.metrics.categoricalCrossentropy(yTrue, model.predict(x));    
    }

    var g = tf.grad(f);
    var grad = g(img);

    return grad;
}
```

The Python equivalent allowed me to succesfully compute the gradient.

```
loss_object = tf.keras.losses.CategoricalCrossentropy(from_logits=True)

//already used one-hot encoding to convert the input_label
def getGradient(input_image, input_label):
  with tf.GradientTape() as tape:
    tape.watch(input_image)
    prediction = model(input_image)
    loss = loss_object(input_label, prediction)

  gradient = tape.gradient(loss, input_image)

  return gradient
```

**Comment by NormanTUD** [comment]
I think I have a very similiar issue. I'm trying to get https://github.com/tensorflow/tfjs-examples/tree/master/visualize-convnet to work, but the gradFunction always returns a zero matrix only.

**Comment by pyu10055** [comment]
@NormanTUD @pxrrylson it would be really helpful if either of you can provide a codepen example that reproduces this problem?  Thanks
